builder/common/step_create_ssm_tunnel.go

```diff
@@ -72,31 +73,61 @@ func (s *StepCreateSSMTunnel) Run(ctx context.Context, state multistep.StateBag)
 	ssmCtx, ssmCancel := context.WithCancel(ctx)
 	s.stopSSMCommand = ssmCancel
 
+	ssmconn := ssm.New(s.AWSSession)
+	session := pssm.Session{
+		SvcClient:  ssmconn,
+		InstanceID: aws.StringValue(instance.InstanceId),
+		RemotePort: s.RemotePortNumber,
+		LocalPort:  s.LocalPortNumber,
+		Region:     s.Region,
+	}
+	go s.CreatePersistentSSMSession(ssmCtx, ui, &session, ssmconn, instance)
+
+	return multistep.ActionContinue
+}
+
+func (s *StepCreateSSMTunnel) CreatePersistentSSMSession(ctx context.Context, ui packersdk.Ui, session *pssm.Session, ssmConn *ssm.SSM, instance *ec2.Instance) {
 	go func() {
-		ssmconn := ssm.New(s.AWSSession)
-		err := pssm.Session{
-			SvcClient:  ssmconn,
-			InstanceID: aws.StringValue(instance.InstanceId),
-			RemotePort: s.RemotePortNumber,
-			LocalPort:  s.LocalPortNumber,
-			Region:     s.Region,
-		}.Start(ssmCtx, ui)
-
-		if err != nil {
-			ui.Error(fmt.Sprintf("ssm error: %s", err))
+		for ctx.Err() == nil {
+			ssmAvailable, err := ssmConn.GetConnectionStatus(&ssm.GetConnectionStatusInput{
```

Comment:
I wonder if we can limit the number of attempts here, from what I understand this will loop without any delay to check the status of the connection.
This will be as responsive as can be from the looks of it, but I think we can probably add some `time.Sleep` to reduce the number of checks we do during a build.

I might be overthinking it though, if you have another idea let me know, always eager to hear suggestions.

Reply:
This is completely valid. I will try it out with different sleep times and see what works with the least number of calls.